### ISSUE TITLE
Print ChainerX internal stack trace

### ISSUE THREAD

**Issue by gwtnb** [chainer/chainer#7420]
ChainerX prints a stack trace when an exception is not caught, but it only contains the python stack. It may difficult to debug only from python stack trace, so printing C++ stack seems preferred.

The following is an example. C++ stack is not shown. It is an artificial example generated by using `LD_PRELOAD`, so it does not indicate there is a bug. Just for a demo.

```
$ LD_PRELOAD=./my_gemm.so python train_mnist.py -d cuda:0
Device: cuda:0
# unit: 1000
# Minibatch-size: 100
# epoch: 20

/path/to/python/lib/python3.5/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
/path/to/chainer/chainer/chainer/_environment_check.py:89: UserWarning: 
--------------------------------------------------------------------------------
Multiple installations of CuPy package has been detected.
You should select only one package from from ['cupy-cuda101', 'cupy-cuda100', 'cupy-cuda92', 'cupy-cuda91', 'cupy-cuda90', 'cupy-cuda80', 'cupy'].
Run `pip list` to see the list of packages currently installed, then
`pip uninstall <package name>` to uninstall unnecessary package(s).
--------------------------------------------------------------------------------

  '''.format(name=name, pkgs=pkgs))
Exception in main training loop: CUBLAS_STATUS_NOT_INITIALIZED
Traceback (most recent call last):
  File "/path/to/chainer/chainer/chainer/training/trainer.py", line 316, in run
    update()
  File "/path/to/chainer/chainer/chainer/training/updaters/standard_updater.py", line 175, in update
    self.update_core()
  File "/path/to/chainer/chainer/chainer/training/updaters/standard_updater.py", line 187, in update_core
    optimizer.update(loss_func, *in_arrays)
  File "/path/to/chainer/chainer/chainer/optimizer.py", line 800, in update
    loss = lossfun(*args, **kwds)
  File "/path/to/chainer/chainer/chainer/link.py", line 294, in __call__
    out = forward(*args, **kwargs)
  File "/path/to/chainer/chainer/chainer/links/model/classifier.py", line 143, in forward
    self.y = self.predictor(*args, **kwargs)
  File "/path/to/chainer/chainer/chainer/link.py", line 294, in __call__
    out = forward(*args, **kwargs)
  File "train_mnist.py", line 27, in forward
    h1 = F.relu(self.l1(x))
  File "/path/to/chainer/chainer/chainer/link.py", line 294, in __call__
    out = forward(*args, **kwargs)
  File "/path/to/chainer/chainer/chainer/links/connection/linear.py", line 155, in forward
    return linear.linear(x, self.W, self.b, n_batch_axes=n_batch_axes)
  File "/path/to/chainer/chainer/chainer/functions/connection/linear.py", line 305, in linear
    y, = LinearFunction().apply(args)
  File "/path/to/chainer/chainer/chainer/function_node.py", line 271, in apply
    outputs = self.forward_chainerx(in_data)
  File "/path/to/chainer/chainer/chainer/functions/connection/linear.py", line 76, in forward_chainerx
    return chainerx.linear(x, W, b),
Will finalize trainer extensions and updater before reraising the exception.
Traceback (most recent call last):
  File "train_mnist.py", line 134, in <module>
    main()
  File "train_mnist.py", line 130, in main
    trainer.run()
  File "/path/to/chainer/chainer/chainer/training/trainer.py", line 349, in run
    six.reraise(*exc_info)
  File "/path/to/python/lib/python3.5/site-packages/six.py", line 693, in reraise
    raise value
  File "/path/to/chainer/chainer/chainer/training/trainer.py", line 316, in run
    update()
  File "/path/to/chainer/chainer/chainer/training/updaters/standard_updater.py", line 175, in update
    self.update_core()
  File "/path/to/chainer/chainer/chainer/training/updaters/standard_updater.py", line 187, in update_core
    optimizer.update(loss_func, *in_arrays)
  File "/path/to/chainer/chainer/chainer/optimizer.py", line 800, in update
    loss = lossfun(*args, **kwds)
  File "/path/to/chainer/chainer/chainer/link.py", line 294, in __call__
    out = forward(*args, **kwargs)
  File "/path/to/chainer/chainer/chainer/links/model/classifier.py", line 143, in forward
    self.y = self.predictor(*args, **kwargs)
  File "/path/to/chainer/chainer/chainer/link.py", line 294, in __call__
    out = forward(*args, **kwargs)
  File "train_mnist.py", line 27, in forward
    h1 = F.relu(self.l1(x))
  File "/path/to/chainer/chainer/chainer/link.py", line 294, in __call__
    out = forward(*args, **kwargs)
  File "/path/to/chainer/chainer/chainer/links/connection/linear.py", line 155, in forward
    return linear.linear(x, self.W, self.b, n_batch_axes=n_batch_axes)
  File "/path/to/chainer/chainer/chainer/functions/connection/linear.py", line 305, in linear
    y, = LinearFunction().apply(args)
  File "/path/to/chainer/chainer/chainer/function_node.py", line 271, in apply
    outputs = self.forward_chainerx(in_data)
  File "/path/to/chainer/chainer/chainer/functions/connection/linear.py", line 76, in forward_chainerx
    return chainerx.linear(x, W, b),
chainerx.ChainerxError: CUBLAS_STATUS_NOT_INITIALIZED
```

**Comment by take-cheeze** [comment]
[abseil](https://abseil.io/about/intro) has stacktracer and symbolizer: https://github.com/abseil/abseil-cpp/tree/master/absl/debugging
And we can replace some third_party submodules with it too.

**Comment by niboshi** [comment]
Related: #7466
It might as well be possible to sleep on exception. (but there may be cases where it is not possible to attach gdb)

**Comment by niboshi** [comment]
Fixed via #7502
